Add RpcAgentTestFixture to extract duplicate code

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29747 Add RpcAgentTestFixture to extract duplicate code**

There are duplicate code for component that rely on RpcAgent. Extract them into a re-usable test fixture class.

Differential Revision: [D5689636](https://our.internmc.facebook.com/intern/diff/D5689636/)

xush6528 committed:
Refactoring torch.distributed.rpc tests to have Mixins and Fixtures

We plan to introduce different `RpcAgentOption`s, this requires us to allow different tests for differnet `RpcAgent`s to correctly configure the test by bringing in different test mixins. This refactoring PR is empowering this.

Also, with more fixtures and mixins, moved tests into sub-folders.

Differential Revision: [D5689636](https://our.internmc.facebook.com/intern/diff/D5689636/)

[ghstack-poisoned]

xush6528 committed:
Update on "Refactoring torch.distributed.rpc tests to have Mixins and Fixtures"


We plan to introduce different `RpcAgentOption`s, this requires us to allow different tests for differnet `RpcAgent`s to correctly configure the test by bringing in different test mixins. This refactoring PR is empowering this.

Also, with more fixtures and mixins, moved tests into sub-folders.

Differential Revision: [D5689636](https://our.internmc.facebook.com/intern/diff/D5689636/)

[ghstack-poisoned]

xush6528 committed:
Update on "Refactoring torch.distributed.rpc tests to have Mixins and Fixtures"


We plan to introduce different `RpcAgentOption`s, this requires us to allow different tests for differnet `RpcAgent`s to correctly configure the test by bringing in different test mixins. This refactoring PR is empowering this.

Also, with more fixtures and mixins, moved tests into sub-folders.

Differential Revision: [D5689636](https://our.internmc.facebook.com/intern/diff/D5689636/)

[ghstack-poisoned]

xush6528 committed:
Update on "Refactoring torch.distributed.rpc tests to have Mixins and Fixtures"


We plan to introduce different `RpcAgentOption`s, this requires us to allow different tests for differnet `RpcAgent`s to correctly configure the test by bringing in different test mixins. This refactoring PR is empowering this.

Also, with more fixtures and mixins, moved tests into sub-folders.

Differential Revision: [D5689636](https://our.internmc.facebook.com/intern/diff/D5689636/)

[ghstack-poisoned]

mrshenli commented on test/run_test.py:
```diff
@@ -370,7 +376,7 @@ def exclude_tests(exclude_list, selected_tests, exclude_message=None):
     for exclude_test in exclude_list:
         for test in tests_copy:
-            if test.startswith(exclude_test):
+            if test == exclude_test:
```
why changing this to exact match?

xush6528 replied:
Notice line 85,

```py
WINDOWS_BLACKLIST = [
    'distributed',  # line 85
    'distributed/rpc/process_group/rpc_fork',
]
```


Initially `selected_tests` is like,
```py
selected_tests = [
    'distributed',
    'distributed/rpc/process_group/rpc_fork',
]
```

After iterating on `exclude_test == 'distributed'`, `selected_tests` becomes,

```py
selected_tests = [
    'distributed/rpc/process_group/rpc_fork',
]
```

Since the `exclude_tests` is,
```py
def exclude_tests(exclude_list, selected_tests, exclude_message=None):
    tests_copy = selected_tests[:]
    for exclude_test in exclude_list:
        for test in tests_copy:
            if test.startswith(exclude_test):
                if exclude_message is not None:
                    print_to_stderr('Excluding {} {}'.format(test, exclude_message))
                selected_tests.remove(test)
    return selected_tests
````

when 
`test == 'distributed/rpc/process_group/rpc_fork'`,
and
`exclude_test == 'distributed'`

It will run `selected_tests.remove('distributed')`, raising a "list does not contain "'distributed'"" error.

mrshenli commented on test/run_test.py:
```diff
@@ -427,17 +433,23 @@ def main():
 
     for test in selected_tests:
-        test_name = 'test_{}'.format(test)
+        splits = test.rsplit("/", 1)
```
jit does this by adding a `test_jit.py` in the `torch/test` folder and then import other tests from there. I wonder if we should do the same.

cc @yf225 @ezyang @suo

mrshenli commented:
I tagged @ezyang because this PR touches `run_test.py` that might affect other modules test. Please let us know if we should keep the same convention as `test_jit.py` and only add minimum changes to `run_test.py`.

mrshenli commented on test/distributed/rpc/process_group/process_group_rpc_agent_mixin.py:
```diff
@@ -0,0 +1,7 @@
+
+class ProcessGroupRpcAgentMixin(object):
+    @property
+    def rpc_backend_name(self):
```
will there be any other helper function be added to this class? If this is the only one, this looks like an overkill for just passing a string to a test class.

xush6528 replied:
Yes, it's `def rpc_agent_options` in the next PR. Adding this Mixin + refactoring for this Mixin is the purpose of this PR.

xush6528 committed:
Update on "Refactoring torch.distributed.rpc tests to have Mixins and Fixtures"


We plan to introduce different `RpcAgentOption`s, this requires us to allow different tests for differnet `RpcAgent`s to correctly configure the test by bringing in different test mixins. This refactoring PR is empowering this.

Also, with more fixtures and mixins, moved tests into sub-folders.

Differential Revision: [D5689636](https://our.internmc.facebook.com/intern/diff/D5689636/)

[ghstack-poisoned]

xush6528 committed:
Update on "Refactoring torch.distributed.rpc tests to have Mixins and Fixtures"


We plan to introduce different `RpcAgentOption`s, this requires us to allow different tests for differnet `RpcAgent`s to correctly configure the test by bringing in different test mixins. This refactoring PR is empowering this.

Also, with more fixtures and mixins, moved tests into sub-folders.

Differential Revision: [D5689636](https://our.internmc.facebook.com/intern/diff/D5689636/)

[ghstack-poisoned]

xush6528 committed:
Update on "Refactoring torch.distributed.rpc tests to have Mixins and Fixtures"


We plan to introduce different `RpcAgentOption`s, this requires us to allow different tests for differnet `RpcAgent`s to correctly configure the test by bringing in different test mixins. This refactoring PR is empowering this.

Also, with more fixtures and mixins, moved tests into sub-folders.

Differential Revision: [D5689636](https://our.internmc.facebook.com/intern/diff/D5689636/)

[ghstack-poisoned]

ezyang commented on test/run_test.py:
```diff
@@ -132,6 +138,15 @@ def print_to_stderr(message):
+    os.environ['PYTHONPATH'] = os.pathsep.join(
+        [path for path in [os.environ.get('PYTHONPATH', ''), test_directory] if path]
+    )
+
```
I think this is silly and you should just keep all the tests in the top level directory. If you do this, will running tests directly from the directory (without using `run_test.py`) work? I don't think so! And now you've just made it that much harder for people to run a specific test with, e.g., gdb attached.

xush6528 replied:
@pietern  Agree? Let's put them flatly?

ezyang reviewed (changes requested):
I'm NACK'ing the `run_test.py` changes. An important invariant I want to hold is that I can run `python test/test_foo.py` without anything special, and I will run that test. The multi-level folder structure that you've setup here causes this property to no longer hold. I don't think multiple folders is worth losing the invariant.

xush6528 committed:
Update on "Refactoring torch.distributed.rpc tests to have Mixins and Fixtures"


We plan to introduce different `RpcAgentOption`s, this requires us to allow different tests for differnet `RpcAgent`s to correctly configure the test by bringing in different test mixins. This refactoring PR is empowering this.

Also, with more fixtures and mixins, moved tests into sub-folders.

Differential Revision: [D5689636](https://our.internmc.facebook.com/intern/diff/D5689636/)

[ghstack-poisoned]

xush6528 committed:
Update on "Refactoring torch.distributed.rpc tests to have Mixins and Fixtures"


We plan to introduce different `RpcAgentOption`s, this requires us to allow different tests for differnet `RpcAgent`s to correctly configure the test by bringing in different test mixins. This refactoring PR is empowering this.

Also, with more fixtures and mixins, moved tests into sub-folders.

Differential Revision: [D5689636](https://our.internmc.facebook.com/intern/diff/D5689636/)

[ghstack-poisoned]

xush6528 commented:
Resolved @ezyang 's comment. Moved all "test_xxxx.py" back to "test/" root directory.

xush6528 commented:
- Replaced `TestConfig` with `RpcAgentTestFixture`. Because, `TestConfig` assumes test config passed from Environment Vars (strings). String type is not sufficient if we want to configure with typed objects like `RpcAgentOptions`.

ezyang commented:
New structure looks fine. Did you want me to review the mixin code too, or are others getting it?

xush6528 commented:
Thanks @ezyang. No other actions needed at the moment.

xush6528 commented:
I have a conclusion with @pritamdamania87 , we don't want to have differnet Test classes for dirrent RPC backends.

We will use trap code to configure 1) which backend the test should us and 2) the RpcAgentOptions to use, so after settingup as a middle step, the test binary still falls to the single test entry class, RpcTest class, DistAutograd class and DistOptimizer class.

xush6528 committed:
Update on "Refactoring torch.distributed.rpc tests to have Mixins and Fixtures"


We plan to introduce different `RpcAgentOption`s, this requires us to allow different tests for differnet `RpcAgent`s to correctly configure the test by bringing in different test mixins. This refactoring PR is empowering this.

Also, with more fixtures and mixins, moved tests into sub-folders.

Differential Revision: [D5689636](https://our.internmc.facebook.com/intern/diff/D5689636/)

[ghstack-poisoned]

mrshenli commented on test/dist_optimizer_test.py:
```diff
@@ -2,13 +2,14 @@
 import torch.distributed.autograd as dist_autograd
 import torch.distributed.rpc as rpc
 import threading
+from rpc_agent_test_fixture import RpcAgentTestFixture
```
move this line up to put it together with dist_utils?

mrshenli commented on test/rpc_test.py:
```diff
@@ -12,14 +12,16 @@
+import dist_utils
+from dist_utils import dist_init
 from torch.distributed.rpc.internal import PythonUDF, _internal_rpc_pickler
+from rpc_agent_test_fixture import RpcAgentTestFixture
```
same here.

mrshenli reviewed (approved):
RpcAgentTestFixture and code refactoring look good to me!

xush6528 committed:
Update on "Add RpcAgentTestFixture to extract duplicate code"


We plan to introduce different `RpcAgentOption`s, this requires us to allow different tests for differnet `RpcAgent`s to correctly configure the test by bringing in different test mixins. This refactoring PR is empowering this.

Also, with more fixtures and mixins, moved tests into sub-folders.

Differential Revision: [D5689636](https://our.internmc.facebook.com/intern/diff/D5689636/)

[ghstack-poisoned]

mrshenli commented:
Shall we remove this line in the PR description as it is no longer relevant?

```
Also, with more fixtures and mixins, moved tests into sub-folders.
```

xush6528 commented:
@mrshenli Removed.

rohan-varma commented:
it looks like this PR broke the windows build? https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-win-ws2016-cuda9-cudnn7-py3-test2/56195/console - (also looks like the culprit from the CI page - https://ezyang.github.io/pytorch-ci-hud/build/pytorch-master)

rohan-varma commented on test/run_test.py:
```diff
@@ -82,20 +83,18 @@
 
 WINDOWS_BLACKLIST = [
     'distributed',
-    'rpc_fork',
```
any reason these are removed from the windows blacklist?

xush6528 replied:
@rohan-varma 
Wanted to include changes in https://github.com/pytorch/pytorch/pull/29827 to avoid that PR rebasing again.
I should have removed line 71, 73 too.

ezyang commented:
For future reference, if you identify that the PR broke a build, you should revert it.
